oxenmq/pubsub.h

```diff
@@ -0,0 +1,135 @@
+#pragma once
+
+#include "connections.h"
+
+#include <unordered_map>
+#include <mutex>
+#include <chrono>
+
+namespace oxenmq
+{
+
+using namespace std::chrono_literals;
+
+struct DummySub {
+    bool operator==(DummySub& other) { return true; }
+};
+
+/**
+ * OMQ Subscription class.  Handles pub/sub connections such that the user
+ * only needs to call methods to subscribe and publish.
+ * FIXME: do we want an unsubscribe, or is expiry / conn management sufficient?
+ *
+ * type T can contain whatever information the user may need at publish time,
+ * for example if the subscription is for logs the subscriber can specify log
+ * levels or categories, and the publisher can choose to send or not based on
+ * those.
+ */
+template <typename T = DummySub>
+class Subscription
+{
+    struct SubData
+    {
+        std::chrono::steady_clock::time_point expiry;
+        T user_data;
+    };
+
+    std::unordered_map<ConnectionID, SubData> subs;
+    std::shared_timed_mutex _mutex;
+    const std::string description; // description of the sub for logging
+    const std::chrono::minutes sub_duration; // extended by re-subscribe
+
+public:
+
+    Subscription() = delete;
+    Subscription(const std::string& description, std::chrono::minutes sub_duration = 30min)
```

Comment:
Take description by value, and std::move into `description`.

Reply:
Also: make this constructor `explicit` because non-explicit single-argument constructors imply implicit conversion.  In this case, not being explicit would mean you can pass a `std::string` as a `Subscription` argument.